L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc

```diff
@@ -21,166 +23,331 @@ class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+      if (frame >= theConfiguration_.maxClustersPerLink() + 2)
+        break;
+      clusters_SA[frame][link] = move(cluster);
+      ++iCluster;
+    }
+    ++iSector60;
   }
 }
 
 void HGCalHistoClusteringWrapper::convertAlgorithmOutputs(
-    const std::vector<l1thgcfirmware::HGCalMulticluster>& multiclusters_out,
-    const std::vector<l1thgcfirmware::HGCalCluster>& rejected_clusters_out,
-    const std::vector<edm::Ptr<l1t::HGCalCluster>>& clustersPtrs,
-    l1t::HGCalMulticlusterBxCollection& multiclustersBXCollection,
-    l1t::HGCalClusterBxCollection& rejected_clusters) const {
-  // Not doing completely the correct thing here
-  // Taking the multiclusters from the stand alone emulation
-  // Getting their consistuent clusters (stand alone objects)
-  // Linking back to the original CMSSW-type cluster
-  // And creating a CMSSW-type multicluster based from these clusters
-  // So the output multiclusters will not be storing bit accurate quantities (or whatever was derived by the stand along emulation)
-  // As these inherit from L1Candidate, could set their HW quantities to the bit accurate ones
-  for (const auto& rejected_cluster : rejected_clusters_out) {
-    rejected_clusters.push_back(0, *clustersPtrs.at(rejected_cluster.index_cmssw()));
-  }
+    l1thgcfirmware::HGCalClusterSAPtrCollection& clusterSums,
+    l1t::HGCalMulticlusterBxCollection& multiClusters_out,
+    const std::vector<std::vector<edm::Ptr<l1t::HGCalCluster>>>& inputClustersPtrs) const {
+  for (const auto& cluster : clusterSums) {
+    // Convert from digitised quantities
+    if (cluster->w() == 0 || cluster->e() == 0)
+      continue;
+    double phi = (cluster->wphi() / cluster->w()) * theConfiguration_.phiRange() / theConfiguration_.phiNValues();
+    double pt = cluster->e() / theConfiguration_.ptDigiFactor();
 
-  std::vector<l1t::HGCalMulticluster> multiclusters;
-  multiclusters.reserve(multiclusters_out.size());
-  for (unsigned int imulticluster = 0; imulticluster < multiclusters_out.size(); ++imulticluster) {
-    bool firstConstituent = true;
-    for (const auto& constituent : multiclusters_out[imulticluster].constituents()) {
-      if (firstConstituent) {
-        multiclusters.emplace_back(clustersPtrs.at(constituent.index_cmssw()), 1.);
-      } else {
-        multiclusters.at(imulticluster).addConstituent(clustersPtrs.at(constituent.index_cmssw()), 1.);
-      }
-      firstConstituent = false;
+    if (pt < 0.5)
+      continue;  // Add (or take) cut threshold to config
+
+    double rOverZ =
+        (cluster->wroz() / cluster->w()) * theConfiguration_.rOverZRange() / theConfiguration_.rOverZNValues();
+    double eta = -1.0 * std::log(tan(atan(rOverZ) / 2));
+    eta *= theConfiguration_.zSide();
+
+    auto sector = theConfiguration_.sector();
+    if (sector == 1) {
+      phi += (2. * M_PI / 3.);
+    } else if (sector == 2) {
+      phi = phi + (4. * M_PI / 3.);
     }
```

Comment:
This seems to be repeated. Might be good to implement the rotation in some function

Reply:
Moved to a function.